src/sentry/auth/providers/saml2.py

```diff
@@ -131,27 +292,61 @@ def retrieve_email(self, attributes, nameid, config):
         ]:
             email_mapping = config['attribute_mapping']['attribute_mapping_email']
             if email_mapping and email_mapping in attributes:
-                return attributes[email_mapping][0]
+                email = attributes[email_mapping][0]
             elif possible_mail:
-                return possible_mail
-            else:
+                email = possible_mail
+            elif identifier:
                 raise Exception(
                     "Email was not provided by the IdP and is required in order to execute the SAML process"
                 )
         elif possible_mail:
-            return possible_mail
-        else:
+            email = possible_mail
+        elif identifier:
             raise Exception("Email mapping is required in order to execute the SAML process")
+        return email
 
-    def retrieve_firstname(self, attributes, config):
-        firstname = None
-        if attributes and 'attribute_mapping' in config and 'attribute_mapping_firstname' in config[
+    def retrieve_username(self, attributes, config, identifier=False):
+        username = None
+        if attributes and 'attribute_mapping' in config and 'attribute_mapping_username' in config[
+            'attribute_mapping'
+        ]:
+            username_mapping = config['attribute_mapping']['attribute_mapping_username']
+            if username_mapping and username_mapping in attributes:
+                username = attributes[username_mapping][0]
+        elif identifier:
+            raise Exception("Username mapping is required in order to execute the SAML process")
+        return username
+
+    def retrieve_displayname(self, attributes, config):
+        displayname = None
+        if attributes and 'attribute_mapping' in config and 'attribute_mapping_displayname' in config[
             'attribute_mapping'
         ]:
-            firstname_mapping = config['attribute_mapping']['attribute_mapping_firstname']
-            if firstname_mapping and firstname_mapping in attributes:
-                firstname = attributes[firstname_mapping][0]
-        return firstname
+            displayname_mapping = config['attribute_mapping']['attribute_mapping_displayname']
+            if displayname_mapping and displayname_mapping in attributes:
+                displayname = attributes[displayname_mapping][0]
+        return displayname
+
+
+class SAML2SLSView(AuthView):
+    def dispatch(self, request, organization_slug):
+        provider = get_provider(organization_slug)
+        if provider is None:
+            raise Http404
+
+        if 'options' not in provider.config or not provider.config['options'].get('options_slo', False):
+            raise Exception("SLO is disabled, can't process that action")
+
+        saml_config = provider.build_saml_config(organization_slug)
+
+        auth = provider.build_auth(request, saml_config)
+        dscb = lambda: logout(request)
```

Comment:
Can you confirm that this request is being made by the current logged in user? Or is this request being sent by the IdP? If it's being sent by the IdP directly, like, from their servers, this won't work. Since there will be no session to logout of.

Either way, I'd also argue that a hard IdP triggered logout should logout all of the user's sessions, and not just the one session. Is it safe to assume that this request would be triggered also when say, someone leaves their company and their account is deactivated? Or would that be a different endpoint?